### PR TITLE
Use containerd/cio

### LINKED ISSUE #483: update containerd for refactoring cio package
https://github.com/containerd/containerd/pull/1895

### PR DESCRIPTION
**Pull request by dnephin**
Closes #483
Blocked on https://github.com/containerd/containerd/pull/1895

With the refactor of `cio.DirectIO` it can be used here.
I also updated vendor to match the changes made in containerd/containerd

### CODE REVIEW

File: vendor.conf [hunk 1]
```diff
@@ -1,7 +1,7 @@
 github.com/blang/semver v3.1.0
 github.com/BurntSushi/toml a368813c5e648fee92e5f6c30e3944ff9d5e8895
 github.com/containerd/cgroups 29da22c6171a4316169f9205ab6c49f59b5b852f
-github.com/containerd/containerd v1.0.0
+github.com/containerd/containerd 85fd6ad246a09426248e2d3345be5a1cb5eb0648 https://github.com/dnephin/containerd
```
[the remark this answers is not in the document]
Comment (mikebrow) [reply vendor.conf]: and to test it out over on a private branch change the repo:
https://github.com/kubernetes-incubator/cri-containerd/blob/master/hack/versions#L4

**Comment by Random-Liu** [comment]
@dnephin We may want to use containerd 1.0.0 for the coming release. Can we do this after the beta release? :)

**Comment by dnephin** [comment]
Yup, no rush. The PR will be here when there is a need to update something else.

**Comment by Random-Liu** [comment]
@dnephin Hi, I'm trying to update containerd to pick up the reaper fix https://github.com/containerd/containerd/commit/90553efdef0678b2609aed74926a487f8ff58d1a.

But need the same refactor here. Do you want to finish this PR? Or do you want me to carry on this? :)

**Comment by dnephin** [comment]
Rebased and updated to use the latest master for the sha

**Comment by Random-Liu** [comment]
Travis test fails:
```
failed to create containerd task: failed to start shim: exec: "containerd-shim": executable file not found in $PATH: unknown
```

**Comment by Random-Liu** [comment]
And:
```
Jan 11 20:54:52 tmp-node-e2e-7ea8ac22-ubuntu-gke-1604-xenial-v20170420-1 cri-containerd[20827]: 2018/01/11 20:54:52 http: panic serving 10.128.0.3:49962: runtime error: invalid memory address or nil pointer dereference
Jan 11 20:54:52 tmp-node-e2e-7ea8ac22-ubuntu-gke-1604-xenial-v20170420-1 cri-containerd[20827]: goroutine 1765 [running]:
Jan 11 20:54:52 tmp-node-e2e-7ea8ac22-ubuntu-gke-1604-xenial-v20170420-1 cri-containerd[20827]: net/http.(*conn).serve.func1(0xc420175c20)
Jan 11 20:54:52 tmp-node-e2e-7ea8ac22-ubuntu-gke-1604-xenial-v20170420-1 cri-containerd[20827]: 	/usr/local/go/src/net/http/server.go:1697 +0xd0
Jan 11 20:54:52 tmp-node-e2e-7ea8ac22-ubuntu-gke-1604-xenial-v20170420-1 cri-containerd[20827]: panic(0x1141440, 0x1c59190)
Jan 11 20:54:52 tmp-node-e2e-7ea8ac22-ubuntu-gke-1604-xenial-v20170420-1 cri-containerd[20827]: 	/usr/local/go/src/runtime/panic.go:491 +0x283
Jan 11 20:54:52 tmp-node-e2e-7ea8ac22-ubuntu-gke-1604-xenial-v20170420-1 cri-containerd[20827]: github.com/containerd/cri-containerd/pkg/server/io.(*ExecIO).Attach(0x0, 0x0, 0x0, 0x1c7b520, 0xc420276120, 0x1c7b520, 0xc420276200, 0x100, 0xc42021bce0, 0x40)
Jan 11 20:54:52 tmp-node-e2e-7ea8ac22-ubuntu-gke-1604-xenial-v20170420-1 cri-containerd[20827]: 	/go/src/github.com/containerd/cri-containerd/pkg/server/io/exec_io.go:61 +0x54
Jan 11 20:54:52 tmp-node-e2e-7ea8ac22-ubuntu-gke-1604-xenial-v20170420-1 cri-containerd[20827]: github.com/containerd/cri-containerd/pkg/server.(*criContainerdService).execInContainer(0xc4202de700, 0x7f25aace37a8, 0xc420480200, 0xc4204835c0, 0x40, 0xc4202765e0, 0x2, 0x2, 0x0, 0x0, ...)
Jan 11 20:54:52 tmp-node-e2e-7ea8ac22-ubuntu-gke-1604-xenial-v20170420-1 cri-containerd[20827]: 	/go/src/github.com/containerd/cri-containerd/pkg/server/container_execsync.go:148 +0x1068
Jan 11 20:54:52 tmp-node-e2e-7ea8ac22-ubuntu-gke-1604-xenial-v20170420-1 cri-containerd[20827]: github.com/containerd/cri-containerd/pkg/server.(*streamRuntime).Exec(0xc42000e420, 0xc420487040, 0x40, 0xc4202765e0, 0x2, 0x2, 0x0, 0x0, 0x1c7b520, 0xc420276120, ...)
Jan 11 20:54:52 tmp-node-e2e-7ea8ac22-ubuntu-gke-1604-xenial-v20170420-1 cri-containerd[20827]: 	/go/src/github.com/containerd/cri-containerd/pkg/server/streaming.go:59 +0x18e
Jan 11 20:54:52 tmp-node-e2e-7ea8ac22-ubuntu-gke-1604-xenial-v20170420-1 cri-containerd[20827]: github.com/containerd/cri-containerd/vendor/k8s.io/kubernetes/pkg/kubelet/server/streaming.(*criAdapter).ExecInContainer(0xc42028de30, 0x0, 0x0, 0x0, 0x0, 0xc420487040, 0x40, 0xc4202765e0, 0x2, 0x2, ...)
Jan 11 20:54:52 tmp-node-e2e-7ea8ac22-ubuntu-gke-1604-xenial-v20170420-1 cri-containerd[20827]: 	/go/src/github.com/containerd/cri-containerd/vendor/k8s.io/kubernetes/pkg/kubelet/server/streaming/server.go:365 +0xf0
Jan 11 20:54:52 tmp-node-e2e-7ea8ac22-ubuntu-gke-1604-xenial-v20170420-1 cri-containerd[20827]: github.com/containerd/cri-containerd/vendor/k8s.io/kubernetes/pkg/kubelet/server/remotecommand.ServeExec(0x1c869a0, 0xc4201b4b60, 0xc4207b2600, 0x1c72920, 0xc42028de30, 0x0, 0x0, 0x0, 0x0, 0xc420487040, ...)
Jan 11 20:54:52 tmp-node-e2e-7ea8ac22-ubuntu-gke-1604-xenial-v20170420-1 cri-containerd[20827]: 	/go/src/github.com/containerd/cri-containerd/vendor/k8s.io/kubernetes/pkg/kubelet/server/remotecommand/exec.go:52 +0x21e
Jan 11 20:54:52 tmp-node-e2e-7ea8ac22-ubuntu-gke-1604-xenial-v20170420-1 cri-containerd[20827]: github.com/containerd/cri-containerd/vendor/k8s.io/kubernetes/pkg/kubelet/server/streaming.(*server).serveExec(0xc42032ebd0, 0xc4207bb830, 0xc4206f7200)
Jan 11 20:54:52 tmp-node-e2e-7ea8ac22-ubuntu-gke-1604-xenial-v20170420-1 cri-containerd[20827]: 	/go/src/github.com/containerd/cri-containerd/vendor/k8s.io/kubernetes/pkg/kubelet/server/streaming/server.go:277 +0x1d4
Jan 11 20:54:52 tmp-node-e2e-7ea8ac22-ubuntu-gke-1604-xenial-v20170420-1 cri-containerd[20827]: github.com/containerd/cri-containerd/vendor/k8s.io/kubernetes/pkg/kubelet/server/streaming.(*server).(github.com/containerd/cri-containerd/vendor/k8s.io/kubernetes/pkg/kubelet/server/streaming.serveExec)-fm(0xc4207bb830, 0xc4206f7200)
Jan 11 20:54:52 tmp-node-e2e-7ea8ac22-ubuntu-gke-1604-xenial-v20170420-1 cri-containerd[20827]: 	/go/src/github.com/containerd/cri-containerd/vendor/k8s.io/kubernetes/pkg/kubelet/server/streaming/server.go:127 +0x3e
Jan 11 20:54:52 tmp-node-e2e-7ea8ac22-ubuntu-gke-1604-xenial-v20170420-1 cri-containerd[20827]: github.com/containerd/cri-containerd/vendor/github.com/emicklei/go-restful.(*Container).dispatch(0xc42032ec60, 0x1c869a0, 0xc4201b4b60, 0xc4207b2600)
Jan 11 20:54:52 tmp-node-e2e-7ea8ac22-ubuntu-gke-1604-xenial-v20170420-1 cri-containerd[20827]: 	/go/src/github.com/containerd/cri-containerd/vendor/github.com/emicklei/go-restful/container.go:277 +0xb18
Jan 11 20:54:52 tmp-node-e2e-7ea8ac22-ubuntu-gke-1604-xenial-v20170420-1 cri-containerd[20827]: github.com/containerd/cri-containerd/vendor/github.com/emicklei/go-restful.(*Container).(github.com/containerd/cri-containerd/vendor/github.com/emicklei/go-restful.dispatch)-fm(0x1c869a0, 0xc4201b4b60, 0xc4207b2600)
Jan 11 20:54:52 tmp-node-e2e-7ea8ac22-ubuntu-gke-1604-xenial-v20170420-1 cri-containerd[20827]: 	/go/src/github.com/containerd/cri-containerd/vendor/github.com/emicklei/go-restful/container.go:120 +0x48
Jan 11 20:54:52 tmp-node-e2e-7ea8ac22-ubuntu-gke-1604-xenial-v20170420-1 cri-containerd[20827]: net/http.HandlerFunc.ServeHTTP(0xc42028dfd0, 0x1c869a0, 0xc4201b4b60, 0xc4207b2600)
Jan 11 20:54:52 tmp-node-e2e-7ea8ac22-ubuntu-gke-1604-xenial-v20170420-1 cri-containerd[20827]: 	/usr/local/go/src/net/http/server.go:1918 +0x44
Jan 11 20:54:52 tmp-node-e2e-7ea8ac22-ubuntu-gke-1604-xenial-v20170420-1 cri-containerd[20827]: net/http.(*ServeMux).ServeHTTP(0xc420289050, 0x1c869a0, 0xc4201b4b60, 0xc4207b2600)
Jan 11 20:54:52 tmp-node-e2e-7ea8ac22-ubuntu-gke-1604-xenial-v20170420-1 cri-containerd[20827]: 	/usr/local/go/src/net/http/server.go:2254 +0x130
Jan 11 20:54:52 tmp-node-e2e-7ea8ac22-ubuntu-gke-1604-xenial-v20170420-1 cri-containerd[20827]: github.com/containerd/cri-containerd/vendor/github.com/emicklei/go-restful.(*Container).ServeHTTP(0xc42032ec60, 0x1c869a0, 0xc4201b4b60, 0xc4207b2600)
Jan 11 20:54:52 tmp-node-e2e-7ea8ac22-ubuntu-gke-1604-xenial-v20170420-1 cri-containerd[20827]: 	/go/src/github.com/containerd/cri-containerd/vendor/github.com/emicklei/go-restful/container.go:292 +0x4d
Jan 11 20:54:52 tmp-node-e2e-7ea8ac22-ubuntu-gke-1604-xenial-v20170420-1 cri-containerd[20827]: net/http.serverHandler.ServeHTTP(0xc42037a410, 0x1c869a0, 0xc4201b4b60, 0xc4207b2600)
Jan 11 20:54:52 tmp-node-e2e-7ea8ac22-ubuntu-gke-1604-xenial-v20170420-1 cri-containerd[20827]: 	/usr/local/go/src/net/http/server.go:2619 +0xb4
Jan 11 20:54:52 tmp-node-e2e-7ea8ac22-ubuntu-gke-1604-xenial-v20170420-1 cri-containerd[20827]: net/http.(*conn).serve(0xc420175c20, 0x1c87be0, 0xc420480e00)
Jan 11 20:54:52 tmp-node-e2e-7ea8ac22-ubuntu-gke-1604-xenial-v20170420-1 cri-containerd[20827]: 	/usr/local/go/src/net/http/server.go:1801 +0x71d
Jan 11 20:54:52 tmp-node-e2e-7ea8ac22-ubuntu-gke-1604-xenial-v20170420-1 cri-containerd[20827]: created by net/http.(*Server).Serve
Jan 11 20:54:52 tmp-node-e2e-7ea8ac22-ubuntu-gke-1604-xenial-v20170420-1 cri-containerd[20827]: 	/usr/local/go/src/net/http/server.go:2720 +0x288
```

**Comment by Random-Liu** [comment]
@dnephin It seems that `make install` didn't install `containerd-shim` in containerd `Makefile`.

File: pkg/server/io/container_io.go [hunk 2]
```diff
@@ -225,21 +213,9 @@ func (c *ContainerIO) Attach(opts AttachOptions) error {
...
-	c.closer.Wait()
-}
-
-// Close closes all FIFOs.
-func (c *ContainerIO) Close() error {
-	c.closer.Close()
-	if c.fifos != nil {
-		return os.RemoveAll(c.fifos.Dir)
+	if c.wg != nil {
+		c.wg.Wait()
```
Comment (Random-Liu) [inline pkg/server/io/container_io.go]: Hm, this doesn't seem correct to me. The `DirectIO.Wait` function seems never called.

Comment (dnephin) [reply pkg/server/io/container_io.go]: `DirectIO.Wait()` is a no-op, so no reason to call it.

**Comment by dnephin** [comment]
The tests seem to be failing with `Received unexpected error rpc error: code = Unavailable desc = grpc: the connection is unavailable`.

I'm not really sure how to debug that failure, or if it has anything to do with the change I made.

**Comment by jterry75** [comment]
@Random-Liu - Should we close stale?

**Pull request closed without merging**